ember-provide-consume-context/src/-private/decorators.ts

```diff
@@ -1,22 +1,33 @@
 import type ContextRegistry from '../context-registry';
-import { DECORATED_PROPERTY_CLASSES } from './provide-consume-context-container';
+import { EMBER_PROVIDE_CONSUME_CONTEXT_KEY } from './provide-consume-context-container';
 import { getContextValue, hasContext } from './utils';
 
 export function provide(contextKey: keyof ContextRegistry) {
   return function decorator(target: any, key: string) {
-    // Track the class as having a decorated property. Later, this will be used
-    // on instances of this component to register them as context providers.
-    const currentContexts = DECORATED_PROPERTY_CLASSES.get(target.constructor);
-    if (currentContexts == null) {
-      DECORATED_PROPERTY_CLASSES.set(target.constructor, {
-        [contextKey]: key,
-      });
-    } else {
-      DECORATED_PROPERTY_CLASSES.set(target.constructor, {
-        ...currentContexts,
-        [contextKey]: key,
-      });
-    }
+    // Define a property on the class, which will later be used on instances of
+    // the component to register them as context providers.
+    const currentContexts = Object.getOwnPropertyDescriptor(
+      target,
+      EMBER_PROVIDE_CONSUME_CONTEXT_KEY,
+    );
+
+    // A class can have multiple @provide decorated properties, we need to
+    // merge the definitions
+    const contextsValue =
+      currentContexts != null
+        ? {
+            ...currentContexts.value,
+            [contextKey]: key,
+          }
+        : {
+            [contextKey]: key,
+          };
```

Comment:
I think we can remove the ternary and simplify it like this instead:

```suggestion
    const contextsValue = {
      ...currentContexts?.value,
      [contextKey]: key,
    };
```

A quick test in firefox shows this works properly:

```
// a = undefined
undefined

b = { ...a, one: 1 }
// Object { one: 1 }

c = {}
// Object {  }

b = { ...a, ...c, one: 1 }
// Object { one: 1 }

b = { ...a, ...c?.test, one: 1 }
// Object { one: 1 }
```